lib/fizzy/instantiate.cpp

```diff
@@ -375,51 +431,38 @@ std::unique_ptr<Instance> instantiate(std::unique_ptr<const Module> module,
 }
 
 std::vector<ExternalFunction> resolve_imported_functions(
-    const Module& module, std::vector<ImportedFunction> imported_functions)
+    const Module& module, const std::vector<ImportedFunction>& imported_functions)
 {
     std::vector<ExternalFunction> external_functions;
     for (const auto& import : module.importsec)
     {
         if (import.kind != ExternalKind::Function)
             continue;
 
-        const auto it = std::find_if(
-            imported_functions.begin(), imported_functions.end(), [&import](const auto& func) {
-                return import.module == func.module && import.name == func.name;
-            });
-
-        if (it == imported_functions.end())
-        {
-            throw instantiate_error{
-                "imported function " + import.module + "." + import.name + " is required"};
-        }
-
         assert(import.desc.function_type_index < module.typesec.size());
         const auto& module_func_type = module.typesec[import.desc.function_type_index];
 
-        if (module_func_type.inputs != it->inputs)
-        {
-            throw instantiate_error{"function " + import.module + "." + import.name +
-                                    " input types don't match imported function in module"};
-        }
-        if (module_func_type.outputs.empty() && it->output.has_value())
-        {
-            throw instantiate_error{"function " + import.module + "." + import.name +
-                                    " has output but is defined void in module"};
-        }
-        if (!module_func_type.outputs.empty() &&
-            (!it->output.has_value() || module_func_type.outputs[0] != *it->output))
-        {
-            throw instantiate_error{"function " + import.module + "." + import.name +
-                                    " output type doesn't match imported function in module"};
-        }
-
-        external_functions.emplace_back(ExternalFunction{it->function, module_func_type});
+        external_functions.emplace_back(find_imported_function(
```

Comment:
Is this refactoring changing behaviour? i.e. it would have stopped with `instantiate_error` in certain cases, but now it will catch an assert?

Reply:
Now, it just changes the order, the checks for `instantiate_error` are moved into `find_imported_function`.

Assertion is guaranteed to be true by parser.

Reply:
Yes I see the checks are moved to `find_imported_function`, but now they are executed post assertions.

Reply:
They are independent - ifs are checking `imported_functions` argument, and assertion checks that `module` is not malformed.